core/src/main/gamelogic/gameState/GameState.kt

```diff
@@ -0,0 +1,26 @@
+package gamelogic.gameState
+
+import Country
+import PositiveInt
+import gamelogic.CountryReinforcement
+import gamelogic.Regrouping
+
+abstract class GameState {
+    open fun addArmies(reinforcements: Collection<CountryReinforcement>): Unit =
```

Comment:
Esto me dejó dudando. Cada uno de los métodos de la interfaz son específicos de alguna subclase (o sea no es una interfaz en el sentido de que todas las subclases la implementan). Quizá en vez de esto se puede pedir al state la acción a realizar,  o sea algo onda `state.perform_action()` y los parametros de la acción deberían ser accesibles desde el state. Igual habría que ver si esto "escala" (esto vaciaria al referee a ser una especie de fachada no mas creo ). Otra opción es agregar algunos ifs en el referee.

Reply:
El tema es que esto no es una interfaz, es una clase abstracta. Las subclases no implementan algunos métodos porque la clase abstracta los implementa. Si esta clase fuese una interfaz, ahí sí todas las subclases implementarían los métodos para que tiren excepción. Y necesitamos que tiren excepción, porque el método makeAttack se puede llamar en cualquier momento.

La única forma de hacer que no tengamos esas excepciones es hacer que Referee tenga un performAction que realice la acción correspondiente dependiendo del momento. Pero ahí entonces:
1. Se pierde declaratividad para el que usa el Referee
2. Que tendría de parámetro? Un objeto que tenga casi todos sus campos en null excepto las cosas que necesita el perform action del momento? Feo feo

O sea, el tema es que no son específicos de cada subclase, porque el makeAttack tiene que hacer algo cuando no estés en el AttackState, aunque sea tirar una excepción. O sea, tirar una excepción no es no hacer nada.

Reply:
Ahí releí y pensé un poco, pero no sé me ocurre como hacerlo lindo lo que decís. Si el state tiene un performAction que accede a los parámetros por medio de fields del referee, entonces muchas veces van a haber fields del referee que tengan valores inútiles o nulls. Y de todas formas van a tener que haber checkeos de si estás en el state de ataque, o sea que la excepción está igual.

Igual entiendo lo que decís, pero yo me convenzo con que esta clase abstracta no es solo una interfaz, sino que también es una implementación por default para sus subclases. O sea, no es que le estamos poniendo que tire una excepción porque en realidad cada state debería implementar una interfaz diferente; los states que no son el AttackState **tienen** que tirar una excepción, porque es parte del contrato del Referee.

Reply:
Sí, el problema de los que digo está en los parámetros. El punto es quién y cómo decide qué se hace. El state pattern nos resuelve el quién. Pero yo me imagino un state pattern más "puro" (no sé si eso existe) en el cual el referee le pregunta qué hacer o le dice "hace lo que haya que hacer" en vez de decir "hacé esto, si podés". 

Creo que el problema acá está en que el que puede validar la acción correcta es el state (para eso está), pero el que conoce los parámetros de la acción es el referee. Por eso para mí era natural que el referee le pregunte al state y después "haga él". Me sigue gustando ese enfoque, previniendo que el referee crezca extrayendo las acciones en clases aparte (o sea en vez de que las acciones sean meros métodos en el referee, hacer abstracciones específicas para ellas tipo `Regroup`, `Reinforce`, `Attack` `Ocuppy` pero no como estados sino como acciones). Y usando pattern matching sobre el tipo del state podemos sacar los ifs.

Entonces, en resumen:

Yo propongo que los states no sepan hacer las acciones pero que el referee tampoco. Que las acciones (reagrupar, atacar, reforzar, ocupar) tengan cada una su clase. Entonces el referee "dispatchea" (está bien usado acá?) de acuerdo al estado. Algo tipo:


`Actuar() { 
    when (state) {
      reinforcement -> Reinforcement(params...).do()
      attack -> when (state.substate) {
                                attack -> Attack(params...).do()
                               ocuppy -> Ocupation(params...).do()
      }
      // ...
   }
}` 

Donde tuve que hacer un sub pattern matching porque el ataque tiene a su vez dos estados.


En fin, solo para debatir un rato y pensar un poco lo que sería mi opinión.Tampoco estoy en contra de tu PR, si no te convence lo que digo decime y te lo apruebo.

Reply:
No lo termino de ver porque el `Referee` tendría que tener sólo el métoedo `Actuar` este que decís en vez de tener `makeAttack` y todos los específicos, porque sino cada uno de los métodos específicos tendría que llamar al `Actuar`, pero ahí como que sería mejor que el método específico maneje la acción en vez de llamar al método `Actuar` que se tiene que fijar en el state actual. Además está el tema de que no sé cómo se pasarían los parámetros. 

Se puede hacer igualmente que el `Referee` ni los states sepan hacer las acciones, y que haya clases aparte para `Occupation` y `Attack` y esas cosas, y ahí habría dos formas de implementarlo: o hacemos que el `Referee` en el `makeAttack` por ejemplo se fije en el state a ver si es attack y que tire o no excepción dependiendo de eso, o hacemos que el state en vez de hacer el ataque él mismo cree un objeto attack y lo devuelva, y que el referee llame a `.run()` en el objeto attack o algo así. Crear el objeto `Attack` permitiría que tengamos un `.undo()` y un `replay()` facilito, porque podemos mantener la lista de todos los commands que se generaron en todo el juego con el resultado del command o cosas así.

Reply:
Si lo de actuar no es bueno y mejor makeAttack, etc.

Me gusta lo de do, redo, etc en las acciones (no se si undo va a valer en kamchatka igual).

Lo de chequear en el referee o mantener la clase abstracta que rompa cuando el estado es incorrecto medio que me da igual a esta altura, solo me parecio muy raro porque en c++ las clases abstractas se usan para definir interfaces, entonces me pareció una 'interfaz' que no era tal.

Dado que está ya hecho así, sigamos con eso y pasemos a lo que sigue